tests/test_algorithm.py

```diff
@@ -1223,6 +1225,416 @@ def test_position_weights(self):
+        prices[::2] = 87
+        prices[1::2] = 100
+        frame.iloc[start_indexes] = np.tile(prices[:, np.newaxis], (1, 5))
+        frame['volume'] = 100
+        frame.fillna(method='ffill', inplace=True)
+
+        return ((sid, frame) for sid in cls.asset_finder.futures_sids)
+
+    def test_expected_shortfall(self):
+        """
+        Test the expected shortfall calculation performed at the end of
+        backtests.
+        """
+        sids = (1, 1004)
+        amounts = (1, 1)
+        calendar = self.trading_calendar
+        af = self.asset_finder
+        equity_1, future_1004 = af.retrieve_all(sids)
+
+        algo = TestPositionWeightsAlgorithm(
+            sids_and_amounts=zip(sids, amounts),
+            sim_params=self.sim_params,
+            env=self.env,
+            benchmark_sid=8554,
+        )
+        daily_stats = algo.run(self.data_portal)
+
+        # Test that we correctly converted the future contract being held into
+        # the appropriate continuous futures.
+        pt_weights = algo.perf_tracker.position_weights
+        future_1000, future_1001, future_1002, future_1003 = af.retrieve_all(
+            [1000, 1001, 1002, 1003],
+        )
+        cf_offset_0 = af.create_continuous_future(
+            root_symbol='CL', offset=0, roll_style='volume', adjustment='mul',
+        )
+        cf_offset_1 = af.create_continuous_future(
+            root_symbol='CL', offset=1, roll_style='volume', adjustment='mul',
+        )
+        cf_offset_2 = af.create_continuous_future(
+            root_symbol='CL', offset=2, roll_style='volume', adjustment='mul',
+        )
+        cf_offset_3 = af.create_continuous_future(
+            root_symbol='CL', offset=3, roll_style='volume', adjustment='mul',
+        )
+        cf_offset_4 = af.create_continuous_future(
+            root_symbol='CL', offset=4, roll_style='volume', adjustment='mul',
+        )
+        day_before_auto_close = calendar.previous_session_label(
+            future_1000.auto_close_date,
+        )
+        self.assertIn(cf_offset_4, pt_weights[day_before_auto_close])
+        self.assertIn(cf_offset_3, pt_weights[future_1000.auto_close_date])
+        day_before_auto_close = calendar.previous_session_label(
+            future_1001.auto_close_date,
+        )
+        self.assertIn(cf_offset_3, pt_weights[day_before_auto_close])
+        self.assertIn(cf_offset_2, pt_weights[future_1001.auto_close_date])
+        day_before_auto_close = calendar.previous_session_label(
+            future_1002.auto_close_date,
+        )
+        self.assertIn(cf_offset_2, pt_weights[day_before_auto_close])
+        self.assertIn(cf_offset_1, pt_weights[future_1002.auto_close_date])
+        day_before_auto_close = calendar.previous_session_label(
+            future_1003.auto_close_date,
+        )
+        self.assertIn(cf_offset_1, pt_weights[day_before_auto_close])
+        self.assertIn(cf_offset_0, pt_weights[future_1003.auto_close_date])
```

Comment:
I had a hard time following the structure in this section.  I'm generally :+1: on adding a little bit of duplication to tests in the interest of being explicit about what's being checked, but in this case I felt like the clarity of the explicitness didn't help because I couldn't see the pattern of what was being checked.